lib/mac_layer/mac.cpp

```diff
@@ -197,35 +283,78 @@ namespace ex2 {
+        // MPDU. That means that none of our tracking variables has been updated,
+        // which means there is not much we can do if, for example, the data
+        // just received was a "middle" MPDU.
 
         // If we already have the first fragment, then continue even if the
-        // header was bad
+        // most recent raw MPDU was bad
         if (m_firstFragmentReceived) {
-          // we did receive a codeword  or codeword fragment, but it's junk since
+          // We did receive a codeword fragment, but it's junk since
           // there was a problem making the MPDU. Increment the count of received
           // MPDUs.
-          m_mpduCount++;
+          m_mpduCodewordFragmentCount++;
           // Maybe this is the final expected MPDU? Better check.
-          if (m_mpduCount == m_expectedMPDUs) {
+          if (m_mpduCodewordFragmentCount == m_numExpectedMpduCodewordFragments) {
+            // We have to pad out the current buffer of received MPDUs and
+            // can't use the MPDU that was supposed to correspond to the just
+            // received data.
+            // Calculate the numMissingMPDUs, which should be 1, but let's be explicit.
+            uint32_t numMissingMPDUs = m_numExpectedMpduCodewordFragments - (m_mpduCodewordFragmentCount - 1);
+            m_codewordBuffer.insert(m_codewordBuffer.end(), numMissingMPDUs * MPDU::maxMTU(), 0);
+
             m_decodePacket();
             return MAC_UHFPacketProcessingStatus::PACKET_READY;
           } // Have all the MPDUs?
+
+          // If it was not the final fragment expected, just continue...
         }
-        else {
-          // If the first user packet fragment has not yet been received, then
-          // we can't tell if this failed MPDU was supposed to be it, so all
-          // we can do is continue to look for a first MPDU
-          return MAC_UHFPacketProcessingStatus::READY_FOR_NEXT_UHF_PACKET;
-        }
-      }
+
+        // If the first user packet fragment has not yet been received, then
+        // we can't tell if this failed MPDU was supposed to be it, so all
+        // we can do is continue to look for a first MPDU; fall through
+
+      } // catch
 
       return MAC_UHFPacketProcessingStatus::READY_FOR_NEXT_UHF_PACKET;
     } // processUHFPacket
 
+    void
+    MAC::m_processFirstMPDU(MPDU &firstMPDU) {
+      // The first MPDU transmitted may have a user packet header in it, which
+      // is necessary at the application layer to make the full packet.
+      // Without it, there is no point in making an application packet so we
+      // keep track of its reception.
+
+      // Assume the first MPDU has correct information in its header.
+      // Save it so we can keep track of the remaining MPDUs needed to
+      // complete the user packet and do some error handling for things
+      // like missing packets.
+      //
+      // @todo what if the header decoded with errors we don't catch?
```

Comment:
Is this todo outdated because of the improved error catching in the header?

Reply:
I am not that confident I've thought of them all, and since I am not doing the live testing, I must leave the question open... But you are right, maybe it's not a useful comment/todo to leave in. May just be a note-to-self :-/

Reply:
I'm sure we'll discover more with more testing